pytket_qir/converter.py

```diff
@@ -0,0 +1,357 @@
+        )
+
+
+class QirConverter:
+    """
+    A class to build and convert the QIR control-flow graph to a pytket circuit.
+    """
+
+    def __init__(
+        self,
+        file_path: str,
+        optimisation_level: int = 0,
+        gateset: Optional[CustomGateSet] = None,
+        wasm_handler: Optional[WasmFileHandler] = None,
+        wasm_int_type: types.Int = types.Int(32),
+        qir_int_type: types.Int = types.Int(64),
+    ) -> None:
+        self.module = QirModule(file_path)
+        self.module_function = self.module.functions[0]
+        self.parser = QirParser(
+            qir_module=self.module,
+            gateset=gateset,
+            wasm_handler=wasm_handler,
+            wasm_int_type=wasm_int_type,
+            qir_int_type=qir_int_type,
+        )
+        self.cfg = OrderedDict()
+        self.rewritten_cfg = self.cfg
+        if optimisation_level == 1:
+            self.collapse_blocks()
+        self.circuit = self.cfg_to_circuit()
+
+    @property
+    def successors(self):
+        return self._successors
+
+    @successors.setter
+    def successors(self, value):
+        self._successors = OrderedDict()
+        for block in self.module.functions[0].blocks:
+            block_name = block.name
+            term = block.terminator
+            # self.conditions[block_name] = True
+
+            # import pdb; pdb.set_trace()
+            if block_name == "entry":
+                # self.conditions["entry_pred"] = True
+                # self.local_conditions["entry_pred"] = True
+                self.edges[block_name] = {block_name: True}
+            if isinstance(term, QirCondBrTerminator):
+                f = lambda el: el.name == term.true_dest or el.name == term.false_dest
+                self._successors[block_name] = list(
+                    map(lambda el: el.name, filter(f, self.module.functions[0].blocks))
+                )
+                self.edges[block_name] = {
+                    term.true_dest: True,
+                    term.false_dest: False,
+                }
+            elif isinstance(term, QirBrTerminator):
+                self._successors[block_name] = [term.dest]
+                self.edges[block_name] = {term.dest: None}
+            elif isinstance(term, QirRetTerminator):
+                self._successors[block_name] = []
+                self.edges[block_name] = {}
+        # Sort the DAG.
+        sorted_dag = topological_sort(self._successors)
+        # Update (sort) the list of successors.
+        for block, succs in sorted_dag.items():
+            if len(succs) > 1:
+                f = (
+                    lambda el: el == self._successors[block][0]
+                    or el == self._successors[block][1]
+                )
+                self._successors[block] = list(filter(f, self._successors))
+
+    @property
+    def predecessors(self):
+        return self._predecessors
+
+    @predecessors.setter
+    def predecessors(self, value):
```

Comment:
Can you maybe add a comment for what this is used?

Reply:
Done.